banzai_floyds/extract.py

```diff
@@ -0,0 +1,190 @@
+
+        pixels_to_bin = np.logical_and(wavelengths >= lower_edge, wavelengths < wavelength_bins[i + 1])
+        # This should be equivalent to Horne 1986 optimal extraction
+        flux = np.sum(weights[pixels_to_bin] * data[pixels_to_bin] * uncertainty[pixels_to_bin]**-2)
+        flux_normalization = np.sum(weights[pixels_to_bin] * uncertainty[pixels_to_bin] ** -2)
+        results['flux'].append(flux / flux_normalization)
+        uncertainty = np.sqrt(np.sum(weights[pixels_to_bin] ** 2 * uncertainty[pixels_to_bin] ** -2))
+        results['fluxerror'].append( uncertainty / flux_normalization)
+    
+    return Table(results)
+
+
+def combine_wavelegnth_bins(wavelength_bins):
+    """
+    Combine wavelength bins, taking the small delta (higher resolution) bins
+    """
+    # Find the overlapping bins
+    # Assume that the orders are basically contiguous and monotonically increasing
+    wavelength_regions = [(min(order_bins['center']), max(order_bins['center'])) for order_bins in wavelength_bins]
+    bin_sizes = [np.mean(order_bins['width']) for order_bins in wavelength_bins]
+
+    # Assume the smaller of the bin widths are from the blue order
+    # We assume here we only have 2 orders and that one order does not fully encompass the other
+    min_wavelength = min(np.array(wavelength_regions).ravel())
+    blue_order_index = 0 if min_wavelength in wavelength_regions[0]['center'] else 1
+    red_order_index = 0 if blue_order_index else 1
+
+    overlap_end_index = np.min(np.argwhere(wavelength_bins[red_order_index]['center'] >
+                                           np.max(wavelength_regions[blue_order_index]['center'])))
+    # clean up the middle partial overlaps
+    middle_bin_upper = wavelength_bins[red_order_index]['center'][overlap_end_index + 1] 
+    middle_bin_upper -= wavelength_bins[red_order_index]['width'][overlap_end_index] / 2.0
+    middle_bin_lower = wavelength_bins[blue_order_index]['center'][-1] + \
+                       wavelength_bins[blue_order_index]['width'] / 2.0
+    middle_bin_center = (middle_bin_upper + middle_bin_lower) / 2.0
+    middle_bin_width = middle_bin_upper - middle_bin_lower
+    overlap_end_index += 1
+    new_bins = {'center': np.hstack([wavelength_bins[blue_order_index]['center'], 
+                                     [middle_bin_center], 
+                                     wavelength_bins[red_order_index][overlap_end_index:]['center']]), 
+                'width': np.hstack([wavelength_bins[blue_order_index]['center'], 
+                                    [middle_bin_width], 
+                                   wavelength_bins[red_order_index][overlap_end_index:]['center']])}
+    return Table(new_bins)
+
+
+class Extractor(Stage):
+    def do_stage(self, image):
+        image.wavelength_bins = get_wavelength_bins(image.wavelengths)
+        profile_centers = fit_profile(image.data, image.uncertainty, image.wavelengths, image.orders,
```

Comment:
a raw FLOYDS science frame doesn't have an uncertainty in it. Will we be adding this somewhere else other than `frames.FLOYDSObservationFrame`?

(There is also no `image.meta`. I believe this is just header information though? I haven't looked closely at the banzai code to check what all it's looking for.)